File: Tracks/csharp/secret-handshake/SecretHandshake.cs
using System;
using System.Collections.Generic;
using System.Linq;

public static class SecretHandshake
{
    private const int ReverseCommand = 1 << 4;
    private static readonly Dictionary<int, string> actions = new Dictionary<int, string>
    {
        {  1 << 0, "wink" },
        {  1 << 1, "double blink" },
        {  1 << 2, "close your eyes" },
        {  1 << 3, "jump" },
    };

    public static string[] Commands(int commandValue)
    {
        bool IncludesCommand(int command) => (commandValue & command) == command;
        var result = actions.Where(commandAction => IncludesCommand(commandAction.Key))
                            .Select(commandAction => commandAction.Value);

        return (IncludesCommand(ReverseCommand) ? result.Reverse() : result).ToArray();
    }
}

File: Tracks/typescript/secret-handshake/secret-handshake.ts
type Command = (input: string[]) => string[];

export default class HandShake {
  private readonly commandsById: Map<number, Command> = new Map([
    [1 << 0, (input: string[]) => input.concat("wink")],
    [1 << 1, (input: string[]) => input.concat("double blink")],
    [1 << 2, (input: string[]) => input.concat("close your eyes")],
    [1 << 3, (input: string[]) => input.concat("jump")],
    [1 << 4, (input: string[]) => input.reverse()],
  ]);

  private readonly commandId: number;

  constructor(commandId: number) {
    this.commandId = commandId;
  }

  commands = (): string[] =>
    [...this.commandsById.entries()].reduce<string[]>(
      (prev, idCommand) =>
        (this.commandId & idCommand[0]) === idCommand[0]
          ? idCommand[1](prev)
          : prev,
      []
    );
}
